src/matching/tagMatcher.ts

```diff
@@ -1,50 +1,88 @@
 export class TagMatcher {
-  static openTag = "<";
-  static closeTag = ">";
-  static closeSlash = "/";
-  static tagNameRegex = /[^\s>]+/;
+  static TAG_REGEX = /\<(\/)?([^\>\<\s]+)[^\>\<]*(\/?)\>/g;
+  static OPEN_FORWARD_SLASH = 1;
+  static TAG_NAME = 2;
+  static CLOSE_FORWARD_SLASH = 3;
 
   openStart: number|undefined;
   openEnd: number|undefined;
   closeStart: number|undefined;
   closeEnd: number|undefined;
 
   constructor(corpus: string, position: number) {
-    let opening = corpus.lastIndexOf(TagMatcher.openTag, position);
-    if (opening === -1) {
-      return;
-    }
+    let match = TagMatcher.TAG_REGEX.exec(corpus);
+    const tags = [];
 
-    // If we found the closing tag, keep searching back for the opening tag
-    if (corpus[opening + 1] === TagMatcher.closeSlash) {
-      position = opening;
-
-      opening = corpus.lastIndexOf(TagMatcher.openTag, opening - 1);
-      if (opening === -1) {
-        return;
+    // Gather all the existing tags.
+    while (match) {
+      // Node is a self closing tag, skip.
+      if (match[TagMatcher.CLOSE_FORWARD_SLASH]) {
+        continue;
       }
-    }
 
-    const tagNameMatch = TagMatcher.tagNameRegex.exec(corpus.substring(opening + 1));
-    if (tagNameMatch === null) {
-      return;
+      tags.push({
+        name: match[TagMatcher.TAG_NAME],
+        type: match[TagMatcher.OPEN_FORWARD_SLASH] ? 'close' : 'open',
+        startPos: match.index,
+        endPos: TagMatcher.TAG_REGEX.lastIndex,
+      });
+
+      match = TagMatcher.TAG_REGEX.exec(corpus);
     }
-    const close = corpus.indexOf(TagMatcher.closeTag, opening + 1);
-    const tagName = tagNameMatch[0];
 
-    // We now know where the opening tag is
-    this.openStart = opening;
-    this.openEnd = close + 1;
+    const stack : any = [];
```

Comment:
Can you give the proper typings here, rather than using `any`? I think it's `{ name: string, type: "close" | "open", startPos: number, endPos: number }[]`